### ISSUE TITLE
value_attribute float being rounded

### ISSUE THREAD

**Issue by amitkeret** [Nerwyn/service-call-tile-feature#29]
**Checklist:**
- [x] I updated to the latest version available
- [x] I cleared the cache of my browser

**Release with the issue:**
3.3.6

**Browser and Operating System:**
Windows 11, Firefox 126

**Description of problem:**
When the value of an entity's attribute is a float, it is being rounded.
Entity states which are floats are returned correctly.

**Your Full Remote Config**
```yaml
# Works as expected:
- type: custom:service-call
    entries:
      - entity_id: sensor.router_upload_speed
        label: "{{ value }}"    # 0.976842....

# Works incorrectly:
- type: custom:service-call
    entries:
      - entity_id: sun.sun
        value_attribute: azimuth
        label: "{{ value }}"    # 281 (Developer console shows this was 281.42)

```

**Comment by Nerwyn** [comment]
It looks like I have [some extra logic](https://github.com/Nerwyn/service-call-tile-feature/blob/3.3.6/src/classes/base-service-call-feature.ts#L463-L469) to truncate values when step/precision isn't defined. For some reason numerical states are returned as strings and numerical attributes are returned as numbers, so this bit of code isn't triggering for states (two wrongs making a right).

I don't think this logic is necessary and I can remove it, but we'll have to do some testing on various services to make sure it doesn't break anything.

**Comment by Nerwyn** [comment]
@amitkeret can you try the 3.3.7 alpha? It should fix the issue.

**Comment by amitkeret** [comment]
testing 3.3.7-beta.007
seems to have two issues here:
- **label**: displays correctly for `button` but not `spinbox` or `slider`
- **using in conditions**: I tried using the entity state/attribute in numeric conditionals. there are three different scenarios here:
  - `{{ value }}`: evaluates as a string and fails the condition. this happens when testing with both state AND attribute
  -  `states(config.entity)` parsed correctly, passes the condition
  - `state_attr(config.entity, [attribute])` also passes the condition, but causes redundancy

I feel like the second issue is part of a broader issue of how `{{ value }}` is being evaluated, but you're a better judge of that :)

```yaml
type: vertical-stack
cards:
  - type: tile
    entity: sensor.upload_speed
    features:
      - type: custom:service-call
        entries:
          - type: button
            label: '{{ value }}'

          # testing conditionals, {{ value }} represents entity state
          - type: button
            color: '{{ iif(value > 0, "green", "blue") }}'  # fails
            label: '{{ value }}'
          - type: button
            color: '{{ iif(states(config.entity) > 0, "green", "blue") }}'  # passes
            label: '{{ value }}'

  - type: tile
    entity: sun.sun
    features:

      # testing label display
      - type: custom:service-call
        entries:
          - type: button
            label: '{{ value }}'
          - type: button
            value_attribute: elevation
            label: '{{ value }}'  # float displays correctly
      - type: custom:service-call
        entries:
          - type: spinbox
            value_attribute: elevation
            label: '{{ value }}'  # float is truncated
          - type: slider
            value_attribute: elevation
            range:
              - 0
              - 100
            label: '{{ value }}'  # float is truncated

      # testing conditionals, {{ value }} represents entity attribute
      - type: custom:service-call
        entries:
          - type: button
            value_attribute: elevation
            color: '{{ iif(value > 30, "green", "blue") }}'  # fails
            label: '{{ value }}'
          - type: spinbox
            value_attribute: elevation
            color: '{{ iif(states(config.entity) > 30, "green", "blue") }}'  # passes the state, not attribute (therefor fails)
            label: '{{ value }}'
          - type: slider
            value_attribute: elevation
            color: '{{ iif(state_attr(config.entity, "elevation") > 30, "green", "blue") }}'  # passes, but is redundant (specifying "elevation" twice in this config
            range:
              - 0
              - 100
            label: '{{ value }}'

```
![image](https://github.com/Nerwyn/service-call-tile-feature/assets/1258597/42898abc-4c75-4653-9442-cddea7711244)

**Comment by Nerwyn** [comment]
Try setting `step` for the slider and spinbox. Their precisions get set by the step size and values get set to that precision to prevent service call errors. Also try applying a numerical filter to value if a service or comparison explicitly requires a number and not a numerical string like `{{ value | int }}` or `{{ value | float }}`. A recent change I made to [ha-nunjucks](https://github.com/Nerwyn/ha-nunjucks) made it so that template result numbers were returned as strings to prevent leading and trailing zeroes from being stripped from numerical strings.

I'm not sure what you mean by that last bullet about `state_attr`.

**Comment by amitkeret** [comment]
@Nerwyn 

- setting the precision fixed the `spinbox` and `slider` label problem 👍 
- adding `int` or `float` filter to templates did not fix the second problem

```yaml
....
  - type: button
    color: '{{ iif(value | float > 0, "green", "blue") }}' # blue
    label: '{{ value }}'
  - type: button
    color: '{{ iif(states(config.entity) > 0, "green", "blue") }}' # green
    label: '{{ value }}'
```

PS my last note re: the use of `state_attr` in a template:
I meant that it's an ok workaround for entity attributes, since the attribute evaluates correctly (as an integer/float and not a string), however there's a duplication there - explicitly typing the attribute name twice, both in `value_attribute` property and in the template itself. ie. this is a potential source for errors, typos etc.

**Comment by Nerwyn** [comment]
`color` is a deprecated field which maps to `style.--color`. The `style` field applies styles to the `service-call-` lit element, which does not have access to internal variables such as `value`. Instead, use internal style fields like `background_style`

```yaml
style: # Does not have access to value and evals to false
  '--color': '{{ iif(value | float > 0, "green", "blue") }}'
background_style: # Has access to value and evals to true
  '--color': '{{ iif(value | float > 0, "green", "blue") }}'
```
![image](https://github.com/Nerwyn/service-call-tile-feature/assets/43071698/4575e8a5-2db1-4910-961f-7957cdb577cb)

As for `state_attr`, you can now access any fields within a custom tile feature entry using `config`. Try using `state_attr(config.entity, config.value_attribute)`

**Comment by Nerwyn** [comment]
Since the original issue of values being incorrectly rounded has been fixed, I'm going to release 3.3.7 and close this bug.

**Comment by Nerwyn** [comment]
Released [3.3.7](https://github.com/Nerwyn/service-call-tile-feature/releases/tag/3.3.7).
